### DISCUSSION TITLE
Parameters as piecewise functions in optimal control

### DISCUSSION THREAD

**Discussion by jimgreenelab** [casadi/casadi#3689] [Q&A]
I am a relatively new user, so I hope you will forgive any ignorance.

I am trying to solve an optimal control problem which minimizes $x(t_{f})$, and is constrained by the ODE

$\dot{x} = \alpha(u)x$,

where $\alpha(u)$ is a piecewise linear function.  I understand how to set this up in CasaADi, and I am trying to use the Opti Stack.  However, I have issues when I call the solver because I defined $\alpha(u)$ via

```python
alpha_u_interpolant = lambda u: casadi.interpolant('alpha_u', 'linear', [u_vals], alpha_data)(u)
``` 
After calling 
```python
 opt.solve()
```
I get a Stack Error of the form

RuntimeError: Error in Opti::solve [OptiNode] at .../casadi/core/optistack.cpp:157:
.../casadi/core/function_internal.cpp:146: Error calling IpoptInterface::init for 'solver':
Error in Function::call for 'nlp' [MXFunction] at .../casadi/core/function.cpp:370:
Error in Function::operator() for 'alpha_u_interpolant' [LinearInterpolant] at .../casadi/core/function.cpp:1522:
.../casadi/core/function_internal.cpp:2013: 'eval_sx' not defined for LinearInterpolant

I suppose what is happening is that somewhere my MX symbolic is being thrown into an SX evaluation.  But I don't fully understand, since the Opti variables are all MX, I think, and thus so is alpha_u_interpolant (if u is MX).  I am no expert here, and I am sure I am missing a subtlety.  

Could someone provide suggestions on the best way to solve such a problem in CasaADi?

**Comment by ernestds** [comment]
Any chance you are calling expand() somewhere or using the expand option?

**Reply by jimgreenelab** [reply]
No.  Not calling expand(), or any expand option.   It's strange because when evaluated on u as an MX symbolic, alpha_u_interpolant(u) claims to be MX.  So not sure why the Opti stack is having problems.
